cluster-autoscaler/cloudprovider/upcloud/upcloud_cloud_provider_test.go

```diff
@@ -0,0 +1,167 @@
+package upcloud
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/upcloud/mocks"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/upcloud/pkg/github.com/upcloudltd/upcloud-go-api/v6/upcloud"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+)
+
+func TestUpCloudCloudProvider_NodeGroups(t *testing.T) {
+	t.Parallel()
+
+	svc := mocks.UpCloudService{}
+	clusterID := uuid.New()
+	p := newUpCloudCloudProvider(clusterID, &svc)
+	require.NoError(t, p.Refresh())
+	svc.NodeGroups = append(svc.NodeGroups, upcloud.KubernetesNodeGroup{Count: 3, Name: "group3"})
+	// node group length should still be 2 as refresh is not yet called
+	require.Len(t, p.NodeGroups(), 2)
+	require.NoError(t, p.Refresh())
+	require.Len(t, p.NodeGroups(), 3)
+}
+
+func TestUpCloudCloudProvider_Name(t *testing.T) {
+	t.Parallel()
+
+	p := UpCloudCloudProvider{}
+	require.Equal(t, cloudprovider.UpCloudProviderName, p.Name())
+}
+
+func TestUpCloudCloudProvider_NodeGroupForNode(t *testing.T) {
+	t.Parallel()
+
+	svc := mocks.UpCloudService{}
+	clusterID := uuid.New()
+	p := newUpCloudCloudProvider(clusterID, &svc)
+	require.NoError(t, p.Refresh())
+
+	group, err := p.NodeGroupForNode(&v1.Node{
+		Spec: v1.NodeSpec{
+			ProviderID: fmt.Sprintf("upcloud:////%s", "group1-1"),
+		},
+	})
+	require.NoError(t, err)
+	require.NotNil(t, group)
+	require.Equal(t, fmt.Sprintf("%s/group1", clusterID.String()), group.Id())
+
+	// nodes of other providers should return `nil` as group and error
+	group, err = p.NodeGroupForNode(&v1.Node{
+		Spec: v1.NodeSpec{
+			ProviderID: fmt.Sprintf("fake:////%s", "group1-1"),
+		},
+	})
+	require.NoError(t, err)
+	require.Nil(t, group)
+}
+
+func TestUpCloudCloudProvider_GetResourceLimiter(t *testing.T) {
+	t.Parallel()
+
+	rl := cloudprovider.NewResourceLimiter(map[string]int64{"min": 1}, nil)
+	p := UpCloudCloudProvider{
+		manager: &manager{
+			clusterID: uuid.New(),
+		},
+		resourceLimiter: rl,
+	}
+	l, err := p.GetResourceLimiter()
+	require.NoError(t, err)
+	require.Equal(t, rl, l)
+}
+
+func TestBuildCloudConfig(t *testing.T) {
```

Comment:
Should we have t.Paraller() here as well?

Reply:
@ajmyyra thanks for the review :+1:  Sadly `t.Paraller()` and `t.Setenv` can't be mixed together.